### PR TITLE
Incremental source generator for actors

### PR DESCRIPTION
**Pull request by m3nax**
# Description

Converted the Actor source generator into an IncrementalSourceGenerator to ease the future definition of new generation pipeline for other objects (E.g: https://github.com/dapr/dotnet-sdk/issues/1305#issuecomment-2234152155)

It also add a null check for actorProxy parameter passed in generated Actor ctor, definition of AnalyzerReleases.Shipped.md and AnalyzerReleases.Unshipped.md diagnostics, remove [deprecated nugets](https://github.com/dotnet/roslyn-sdk/blob/main/src/Microsoft.CodeAnalysis.Testing/README.md#obsolete-packages).

## Issue reference

_

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [ ] Extended the documentation

### CODE REVIEW

File: all.sln
Comment (KrylixZA) [inline all.sln:350]: Something looks wrong here. It should at least say `GlobalSection`, but are you sure you need this specific section of changes at all?

File: src/Dapr.Actors.Generators/ActorClientGenerator.cs
Comment (KrylixZA) [inline src/Dapr.Actors.Generators/ActorClientGenerator.cs:60]: Meaningful documentation around parameters, expected returns and exceptions thrown goes a long way to making the library easier to use. Please could you fill in the docs here?

Comment (m3nax) [reply src/Dapr.Actors.Generators/ActorClientGenerator.cs:60]: Tomorrow i will improve summaries.

File: src/Dapr.Actors.Generators/Extensions/IEnumerableExtensions.cs
Comment (KrylixZA) [inline src/Dapr.Actors.Generators/Extensions/IEnumerableExtensions.cs:8]: Please remember to document the params here too.

Comment (KrylixZA) [reply src/Dapr.Actors.Generators/Extensions/IEnumerableExtensions.cs:8]: Maybe just double check all XML docs for this PR.

File: all.sln
Comment (m3nax) [reply all.sln]: Fixed

**Comment by m3nax** [comment]
@KrylixZA really thank for taking time to review my pr.

File: src/Dapr.Actors.Generators/ActorClientGenerator.cs
Comment (KrylixZA) [inline src/Dapr.Actors.Generators/ActorClientGenerator.cs:63]: I presume the omission of an access modifier to this method is deliberate? Could we get away with this being private, or does it need to be internal (C# default)?

Same question for the `GenerateActorClientCode` method below.

File: src/Dapr.Actors.Generators/Templates.cs
Comment (KrylixZA) [inline src/Dapr.Actors.Generators/Templates.cs:21]: I am not going to dictate here, but I do prefer to use `is null` rather than `== null` checks in my code. This is because the `==` can be overwritten to behave differently. It's highly unlikely to happen, but it can. You can read up about it [on this StackOverflow thread](https://stackoverflow.com/questions/40676426/what-is-the-difference-between-x-is-null-and-x-null) if you'd like. 

If that's not your style, I'm happy for you to resolve this comment with a simple comment saying something like "I'll stick to `== null` in my code".

**Review by KrylixZA — APPROVED** [review]
My knowledge of source generators is limited. I joined the review more to learn than anything else. Thanks for the education! 😉

That said, LGTM. We'll still need one of the Dapr maintainers to approve before it merges though as I am not on that list yet.

Thanks for your contributing!

File: src/Dapr.Actors.Generators/ActorClientGenerator.cs
Comment (m3nax) [reply src/Dapr.Actors.Generators/ActorClientGenerator.cs]: Fixed

File: src/Dapr.Actors.Generators/Templates.cs
Comment (m3nax) [reply src/Dapr.Actors.Generators/Templates.cs]: Handled, thank

**Comment by m3nax** [comment]
@WhitWaldo I have resolved the conflicts with the Master, do you have time to look at this branch?

File: src/Dapr.Actors.Generators/Extensions/IEnumerableExtensions.cs
Comment (WhitWaldo) [inline src/Dapr.Actors.Generators/Extensions/IEnumerableExtensions.cs:21]: Instead of a foreach with a separate variable tracking the index, why not just use a for and repurpose the incrementing value?

File: src/Dapr.Actors.Generators/Helpers/SyntaxFactoryHelpers.cs
Comment (WhitWaldo) [inline src/Dapr.Actors.Generators/Helpers/SyntaxFactoryHelpers.cs:10]: Do these need to be public or can they just be internal?

**Review by WhitWaldo — COMMENTED** [review]
Two nits to address - otherwise looks good.

File: src/Dapr.Actors.Generators/Helpers/SyntaxFactoryHelpers.cs
Comment (m3nax) [reply src/Dapr.Actors.Generators/Helpers/SyntaxFactoryHelpers.cs]: You are right

File: src/Dapr.Actors.Generators/Extensions/IEnumerableExtensions.cs
Comment (m3nax) [reply src/Dapr.Actors.Generators/Extensions/IEnumerableExtensions.cs:21]: To avoid enumerating the source parameter twice (one for the Count() and one to iterate source elements), I do not use the `Count()` function that I need to write the `for(int i=0; i < source.Count(); i++)`

**Review by WhitWaldo — APPROVED** [review]
Looks good to me!

**Pull request merged**
